source/common/grpc/async_client_impl.h

```diff
@@ -29,54 +31,76 @@ class AsyncClientImpl final : public AsyncClient<RequestType, ResponseType> {
   }
 
   // Grpc::AsyncClient
-  AsyncClientStream<RequestType>*
-  start(const Protobuf::MethodDescriptor& service_method,
-        AsyncClientCallbacks<ResponseType>& callbacks,
-        const Optional<std::chrono::milliseconds>& timeout) override {
-    std::unique_ptr<AsyncClientStreamImpl<RequestType, ResponseType>> grpc_stream{
-        new AsyncClientStreamImpl<RequestType, ResponseType>(*this, callbacks)};
-    Http::AsyncClient::Stream* http_stream =
-        cm_.httpAsyncClientForCluster(remote_cluster_name_)
-            .start(*grpc_stream, Optional<std::chrono::milliseconds>(timeout));
-
-    if (http_stream == nullptr) {
-      callbacks.onRemoteClose(Status::GrpcStatus::Unavailable);
+  AsyncRequest* send(const Protobuf::MethodDescriptor& service_method, const RequestType& request,
+                     AsyncRequestCallbacks<ResponseType>& callbacks,
+                     const Optional<std::chrono::milliseconds>& timeout) override {
+    std::unique_ptr<AsyncStreamImpl<RequestType, ResponseType>> grpc_stream{
+        new AsyncRequestImpl<RequestType, ResponseType>(*this, service_method, request, callbacks,
+                                                        timeout)};
+
+    grpc_stream->initialize();
+    if (grpc_stream->hasResetStream()) {
       return nullptr;
     }
 
-    grpc_stream->set_stream(http_stream);
+    grpc_stream->moveIntoList(std::move(grpc_stream), active_streams_);
+    return dynamic_cast<AsyncRequestImpl<RequestType, ResponseType>*>(
```

Comment:
perf nit: This has cost. Typically I would just hold a local reference/pointer to the unique_ptr, then return it after you add to the list.